host/sgx/ocalls.c

```diff
@@ -41,28 +41,32 @@ void HandleThreadWait(oe_enclave_t* enclave, uint64_t arg_in)
 
 #if defined(__linux__)
 
-    if (__sync_fetch_and_add(&event->value, (uint32_t)-1) == 0)
+    if (__sync_fetch_and_add(event, (uint32_t)-1) == 0)
     {
         do
         {
-            syscall(
-                __NR_futex,
-                &event->value,
-                FUTEX_WAIT_PRIVATE,
-                -1,
-                NULL,
-                NULL,
-                0);
+            syscall(__NR_futex, event, FUTEX_WAIT_PRIVATE, -1, NULL, NULL, 0);
             // If event->value is still -1, then this is a spurious-wake.
             // Spurious-wakes are ignored by going back to FUTEX_WAIT.
             // Since FUTEX_WAIT uses atomic instructions to load event->value,
             // it is safe to use a non-atomic operation here.
-        } while (event->value == (uint32_t)-1);
+        } while (*event == (uint32_t)-1);
     }
 
 #elif defined(_WIN32)
 
-    WaitForSingleObject(event->handle, INFINITE);
+    uint32_t LOCK_IS_FREE = 0;
+    uint32_t LOCK_IS_TAKEN = 1;
+    // Change the event to LOCK_IS_TAKEN from LOCK_IS_FREE.
+    if (_InterlockedCompareExchange(event, LOCK_IS_TAKEN, LOCK_IS_FREE) == LOCK_IS_FREE)
+    {
+        // Waiting while event is LOCK_IS_TAKEN.
+        // Add a loop in case of fake waking.
+        do
+        {
+            WaitOnAddress(event, &LOCK_IS_TAKEN, sizeof(*event), INFINITE);
```

Comment:
@nonpolarity  I think this should be WaitOnAddress(event, &LOCK_IS_FREE, sizeof(*event), INFINITE). I might be wrong but https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitonaddress says the second argument should be "A pointer to the location of the previously observed value at Address". 


UINT32 myVar = TRUE;

void Wait()
{
        UINT32 unDesiredValue = FALSE;
        InterlockedExchange(&myVar, unDesiredValue);        
         WaitOnAddress(&myVar, &unDesiredValue, sizeof(unDesiredValue), INFINITE);


 }

void SetEvent()
{
    if (!myVar)
    {

        InterlockedExchange(&myVar, TRUE);
        WakeByAddressAll((PVOID)&myVar);
    }
}

Reply:
Thanks @radhikaj , I will check and test.

Reply:
@nonpolarity, I read the code wrong. Before the wait the value on event is lock-taken. So your parameters are correct. Ignore previous comment